controllers/flinkcluster_reconciler.go

```diff
@@ -520,6 +537,49 @@ func (reconciler *ClusterReconciler) cancelFlinkJob(jobID string, takeSavepoint
 	return reconciler.flinkClient.StopJob(apiBaseURL, jobID)
 }
 
+// Trigger savepoint if it is possible, then return the savepoint status to update.
+// When savepoint was already triggered, return the current observed savepoint.
+// If triggering savepoint is skipped or triggered savepoint was created, proceed to stop the job.
+func (reconciler *ClusterReconciler) cancelFlinkJobAsync(jobID string, takeSavepoint bool) (*v1beta1.SavepointStatus, error) {
```

Comment:
To make it more readable, can we refactor this function into a state machine like:

```
currentState = getCurrentState(...)
switch(currentState) {
  case savepointNotTriggered:
    triggerSavepoint()
  case savepointNotFinished:
    // wait for the next iteration
  case savepointSuceeded:
    ...
  case savepointFailed:
    ...
}
```

Reply:
Thanks for the advice. I've improved savepoint status.